### PR TITLE
Migrate PBS from using embedded Vert.x to verticles

### PR DESCRIPTION
**Pull request by And1sS**
To bump Vert.x to version 4 we need to move from embedded usage (via [ContextRunner](https://github.com/prebid/prebid-server-java/blob/master/src/main/java/org/prebid/server/vertx/ContextRunner.java) class) to Vert.x verticles deployment model usage. This is due to updates in Vert.x internal context management. 

Previously, we were obtaining new event loop contexts via `vertx.getOrCreateContext()` calls from spring boot config classes. When spring boot bootstraps application context it runs all configurations on single thread. Since Vert.x 4, call to `vertx.getOrCreateContext()` is going to create new event loop context for the first time and put it in thread local variable, so subsequent calls will just return same event loop context. It breaks [ContextRunner](https://github.com/prebid/prebid-server-java/blob/master/src/main/java/org/prebid/server/vertx/ContextRunner.java) class) and our deployment model.

So, we decided to switch to verticles deployment model, where each instance of deployed verticle is assigned new event loop context and that is guaranteed by contract. This way we are not relying on vertx internal implementation.

### CODE REVIEW

File: src/main/java/org/prebid/server/handler/OptoutHandler.java [hunk 1]
```diff
@@ -41,6 +44,13 @@ public OptoutHandler(GoogleRecaptchaVerifier googleRecaptchaVerifier, UidsCookie
         this.optinUrl = Objects.requireNonNull(optinUrl);
     }
 
+    @Override
+    public List<HttpEndpoint> endpoints() {
+        return List.of(
+                HttpEndpoint.of(HttpMethod.GET, "/optout"),
+                HttpEndpoint.of(HttpMethod.POST, "/optout"));
+    }
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/handler/OptoutHandler.java:52]: I don't like the idea that handler, can have more then one endpoint.
It makes a thought that single responsibility principle is broken.
Maybe it is worth to separate for two different? And make ApllicationResource class have only `endpoint` method?

File: src/main/java/org/prebid/server/vertx/verticles/VerticleDeployer.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+import java.util.Objects;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class VerticleDeployer {
+
+    private static final int TIMEOUT_MILLIS = 5000;
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/vertx/verticles/VerticleDeployer.java:14]: To application properties?

File: src/main/java/org/prebid/server/vertx/verticles/server/ServerVerticle.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                .requestHandler(router);
+
+        if (exceptionHandler != null) {
+            server.exceptionHandler(exceptionHandler);
+        }
+
+        server.listen(address, this::onServerStarted);
+    }
+
+    private void onServerStarted(AsyncResult<HttpServer> result) {
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/vertx/verticles/server/ServerVerticle.java:63]: onServerStart

File: src/main/java/org/prebid/server/vertx/verticles/server/application/ApplicationResource.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package org.prebid.server.vertx.verticles.server.application;
+
+import io.vertx.core.Handler;
+import io.vertx.ext.web.RoutingContext;
+import org.prebid.server.vertx.verticles.server.HttpEndpoint;
+
+import java.util.List;
+
+public interface ApplicationResource extends Handler<RoutingContext> {
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/vertx/verticles/server/application/ApplicationResource.java:9]: HttpResource extends ApplicationResource?
And use HttpResource for handlers?

File: src/main/resources/application.yaml [hunk 5]
```diff
@@ -19,14 +19,16 @@ server:
   http:
     enabled: true
     port: 8080
-    server-instances: 1
+    server-instances: 10
```
Comment (SerhiiNahornyi) [inline src/main/resources/application.yaml:22]: Why 10?

File: src/main/resources/application.yaml [hunk 6]
```diff
@@ -19,14 +19,16 @@ server:
...
 admin:
   port: 8060
 admin-endpoints:
+  credentials:
+    admin: admin
   version:
-    enabled: false
+    enabled: true
```
Comment (SerhiiNahornyi) [inline src/main/resources/application.yaml:29]: Why changed?

**Review by SerhiiNahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/spring/config/server/admin/AdminServerAuthProvider.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        if (MapUtils.isEmpty(credentials)) {
+            resultHandler.handle(Future.failedFuture("Credentials not set in configuration."));
+            return;
+        }
+
+        final String requestUsername = authInfo.getString("username");
+        final String requestPassword = StringUtils.chomp(authInfo.getString("password"));
+
+        final String storedPassword = credentials.get(requestUsername);
+        if (StringUtils.isNotBlank(requestPassword) && Objects.equals(storedPassword, requestPassword)) {
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/spring/config/server/admin/AdminServerAuthProvider.java:34]: StringUtils.equesl

**Review by SerhiiNahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/spring/config/metrics/PrometheusConfiguration.java [hunk 8]
```diff
@@ -44,51 +62,22 @@ public SampleBuilder sampleBuilder(PrometheusConfigurationProperties prometheusC
...
-
-            contextRunner.<HttpServer>runOnServiceContext(promise ->
-                    vertx.createHttpServer()
-                            .requestHandler(router)
-                            .listen(prometheusConfigurationProperties.getPort(), promise));
-
-            logger.info("Successfully started Prometheus Server");
-        }
+            return router;
+        };
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/spring/config/metrics/PrometheusConfiguration.java:80]: What about to wrap `MetricsHandler` with `ApplicationResource` implementation?

File: src/main/java/org/prebid/server/spring/config/server/application/ApplicationServerConfiguration.java [hunk 9]
```diff
@@ -120,48 +154,35 @@ ExceptionHandler exceptionHandler(Metrics metrics) {
...
-                  StaticHandler staticHandler) {
+    @Bean
+    BiFunction<String, SocketAddress, Verticle> applicationVerticleFactory(
+            @Qualifier("applicationServerRouter") Router router,
+            HttpServerOptions httpServerOptions,
+            ExceptionHandler exceptionHandler) {
+
+        return (name, address) -> new ServerVerticle(
+                name, httpServerOptions, address, router, exceptionHandler);
+    }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/spring/config/server/application/ApplicationServerConfiguration.java:165]: I would get rid of this factory. It feels too redundant for me.

File: src/main/java/org/prebid/server/spring/config/server/application/ApplicationServerConfiguration.java [hunk 10]
```diff
@@ -120,48 +154,35 @@ ExceptionHandler exceptionHandler(Metrics metrics) {
...
+
+    @Bean("applicationServerRouter")
+    Router applicationServerRouter(
+            BodyHandler bodyHandler,
+            NoCacheHandler noCacheHandler,
+            CorsHandler corsHandler,
+            List<ApplicationResource> resources,
+            @Qualifier("applicationPortAdminResourcesBinder") AdminResourcesBinder adminResourcesBinder,
+            StaticHandler staticHandler) {
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/spring/config/server/application/ApplicationServerConfiguration.java:174]: I don't see any other use of Vertx in this file, let's inject it directly to create this bean.

File: src/main/java/org/prebid/server/vertx/verticles/server/ServerVerticle.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        this.name = Objects.requireNonNull(name);
+        this.serverOptions = null;
+        this.address = Objects.requireNonNull(address);
+        this.router = Objects.requireNonNull(router);
+        this.exceptionHandler = null;
+    }
+
+    @Override
+    public void start(Promise<Void> startPromise) {
+        final HttpServerOptions httpServerOptions = ObjectUtils.defaultIfNull(serverOptions, new HttpServerOptions());
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/vertx/verticles/server/ServerVerticle.java:50]: `ObjectUtils.getIfNull(serverOptions, HttpServerOptions::new);`

**Review by CTMBNara — CHANGES_REQUESTED** [review]

**Review by Net-burst — APPROVED** [review]
LGTM

**Pull request merged**